pkg/sql/copy/copy_test: TestLargeCopy failed

pkg/sql/copy/copy_test.TestLargeCopy [failed](https://mesolite.cluster.engflow.com/invocations/default/032cc338-3da1-4473-bdae-0998e9571e01?testReportRun=1&testReportShard=3&testReportAttempt=1#targets-Ly9wa2cvc3FsL2NvcHk6Y29weV90ZXN0) on master @ [7d7551f6cf7bd8078139a3c377dd1db1ddf2480e](https://github.com/cockroachdb/cockroach/commits/7d7551f6cf7bd8078139a3c377dd1db1ddf2480e):


```
=== RUN   TestLargeCopy
    test_log_scope.go:171: test logs captured to: outputs.zip/logTestLargeCopy3364558056
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:182: automatically injected an external process virtual cluster under test; see comment at top of test_server_shim.go for details.
    copy_test.go:767: ERROR: restart transaction: TransactionRetryWithProtoRefreshError: TransactionAbortedError(ABORT_REASON_CLIENT_REJECT): "sql txn" meta={id=ff6778c1 key=/Tenant/10/Table/104/1/0/-6774234413454873929/0 iso=Serializable pri=0.00218810 epo=0 ts=1767690357.214826815,2 min=1767690348.805216357,0 seq=17856} lock=true stat=PENDING rts=1767690348.805216357,0 gul=1767690349.305216357,0 obs={n1@1767690348.931424212,0} (SQLSTATE 40001)
    panic.go:615: -- test log scope end --
test logs left over in: outputs.zip/logTestLargeCopy3364558056
--- FAIL: TestLargeCopy (17.06s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=1</code>
 - <code>shard=3</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLargeCopy.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-58406

--- a/pkg/sql/copy/copy_test.go
+++ b/pkg/sql/copy/copy_test.go
@@ -723,9 +723,11 @@ func TestLargeCopy(t *testing.T) {
 	rows := rng.Intn(2000) + 1
 	cr := &copyReader{rng: rng, cols: desc.PublicColumns(), rows: rows}
 
+	atomicCopy := true
 	if rng.Float64() < 0.5 {
 		_, err = sqlDB.Exec("SET copy_from_atomic_enabled = false")
 		require.NoError(t, err)
+		atomicCopy = false
 	}
 
 	// In 50% cases change the primary key concurrently with the COPY.
@@ -734,14 +736,16 @@ func TestLargeCopy(t *testing.T) {
 		if err == nil {
 			return true
 		}
+		if atomicCopy && strings.Contains(err.Error(), "restart transaction") {
+			// We can only retry txn errors with non-atomic COPY.
+			return true
+		}
 		if !doAlterPK {
 			return false
 		}
 		// We might hit a duplicate key error when changing the primary key
 		// (which seems somewhat expected), so we'll ignore such an error.
 		return strings.Contains(err.Error(), "duplicate key") ||
-			// We might be forced to restart - also seems somewhat expected.
-			strings.Contains(err.Error(), "restart transaction") ||
 			// TODO(yuzefovich): occasionally we get this error, and it's not
 			// clear why. Look into this.
 			strings.Contains(err.Error(), "cannot disable pipelining on a running transaction")
@@ -761,6 +765,9 @@ func TestLargeCopy(t *testing.T) {
 	} else {
 		close(alterPKCh)
 	}
+	// TODO(yuzefovich): if we have atomic COPY and get a txn retry error, we
+	// could consider explicitly reruning the COPY command a few times rather
+	// than just ignoring the error.
 	numrows, copyErr := conn.GetDriverConn().CopyFrom(ctx, cr, "COPY lineitem FROM STDIN WITH CSV;")
 	alterErr := <-alterPKCh
 	if !ignoreErr(copyErr) {

